### PR TITLE
New mesh door models, and extensive door API

### PR DESCRIPTION
**Pull request by sofar**
This patch replaces the default door nodes with a new mesh model
and nodes.

Two new models were added that are 2 blocks high. One for left-hinge
and one for right-hinge doors. This allows us to make a single texture
fit on both models. The alternative would have been 1 model and 2
unmapped textures, which is more work for mod developers.

Doors work exactly like the old doors, including ownership, breaking
doors, opening and closing.

Under the hood, we can prevent the top part of the door from being
obstructed by placing an invisible node. This prevents liquids from
flowing through doors or people placing sand or other blocks in the
top half. The door code automatically places and removes these as
needed.

Metadata is used to store door state, just like the old version.

A doors API is added, it allows mods to use the API to open/close or
toggle door states without worrying about sounds, permissions and
other details. This is intended for e.g. mesecons. This API allows
mods to manipulate or inspect doors for players or for themselves.

In-game old door nodes are automatically converted using an ABM and
preserve ownership and orientation and state.

TNT blows up all doors and trapdoors except for the steel ones,
who can survive a blast.

Thanks to red-001 red-001@users.noreply.github.com for some
of the conversion code, cleanups, and extra textures.

### CODE REVIEW

**Comment by MoNTE48** [comment]
-0.5, this broken compatibility with Minecraft and Minetest texture packs...

**Comment by tenplus1** [comment]
+1

**Comment by Jordach** [comment]
It would take less than two minutes to merge the textures into the UV layout. Not really that difficult with GIMP or Photoshop.

-----Original Message-----

**Comment by sofar** [comment]
@MoNTE48 The UV layout is really easy to understand, and allows for more complex door textures than before, including different door inside/outside textures.

File: mods/doors/README.txt [hunk 1]
```diff
@@ -40,8 +43,20 @@ Steel trapdoor textures by sofar (CC-BY-SA-3.0)
   doors_trapdoor_steel.png
   doors_trapdoor_steel_side.png
 
+Door 3d models by sofar (CC-BY-SA-3.0)
+  door_a.obj
+  door_b.ob
```
Comment (kaeza) [inline mods/doors/README.txt]: j?

File: mods/doors/init.lua [hunk 2]
```diff
@@ -1,435 +1,440 @@
...
-	end
-	if not def.selection_box_bottom then
-		def.selection_box_bottom= box
+--]]
+
+-- our API object
+doors = {}
+
+-- private data
+_doors = {}
```
Comment (kaeza) [inline mods/doors/init.lua]: If it's private, please make local.

File: mods/doors/init.lua [hunk 3]
```diff
@@ -1,435 +1,440 @@
...
+_doors.registered_doors = {}
+_doors.registered_trapdoors = {}
+
+-- returns an object to a door object or nil
+doors.get = function(pos)
```
Comment (kaeza) [inline mods/doors/init.lua]: Minor nitpicking, but can the `foo.bar = function(...)` form just die already? `function foo.bar(...)` does exactly the same and looks prettier <3

File: mods/doors/init.lua [hunk 4]
```diff
@@ -1,435 +1,440 @@
...
+	if state % 2 == 1 then
+		state = state - 1
+	else
+		state = state + 1
 	end
-	if not def.sound_open_door then
-		def.sound_open_door = "doors_door_open"
+
+	local dir = minetest.get_node(pos).param2
+	local transform = {
```
Comment (kaeza) [inline mods/doors/init.lua]: This could be moved outside the function so it's initialized only once.

File: mods/doors/init.lua [hunk 5]
```diff
@@ -1,435 +1,440 @@
...
+		minetest.sound_play(def.sound_close, {pos = pos, gain = 0.3, max_hear_distance = 10})
+		minetest.swap_node(pos, {name = string.sub(node.name, 1, string.len(node.name) - 5), param1 = node.param1, param2 = node.param2})
+	else
+		minetest.sound_play(def.sound_open, {pos = pos, gain = 0.3, max_hear_distance = 10})
+		minetest.swap_node(pos, {name = node.name .. "_open", param1 = node.param1, param2 = node.param2})
+	end
+end
 
-function doors.register_trapdoor(name, def)
+doors.register_trapdoor = function(name, def)
```
Comment (kaeza) [inline mods/doors/init.lua]: Why?

File: mods/doors/init.lua (hunk 2 above)
Comment (sofar) [inline mods/doors/init.lua]: Good one, yes, this one is supposed to be ```local```

File: mods/doors/init.lua (hunk 3 above)
Comment (sofar) [inline mods/doors/init.lua]: sorry, I'm happy with either form, but unless there's a style guide, there will just be someone else nitpicking the other way.

I'm not opposed to your change, this file originally mixed both forms, hence me changing all the forms to the same style, which is an improvement.

File: mods/doors/init.lua (hunk 4 above)
Comment (sofar) [inline mods/doors/init.lua]: sure, will do

**Comment by sofar** [comment]
@kaeza pushed an update addressing all your comments, including the ```function foo()``` change.

File: mods/doors/README.txt (hunk 1 above)
Comment (sofar) [inline mods/doors/README.txt]: ah, lol, didn't see that.

File: mods/doors/init.lua (hunk 3 above)
Comment (kaeza) [inline mods/doors/init.lua]: I guess it's like the eternal war about where to put the braces in C-like languages.

This is just personal opinion, but I (generally) use `function foo.bar()` when I mean to use `foo` as a namespace, while using `foo.bar = function()` if `foo` is an object/record/whatever. Sort of implicit documentation or something.

Did not mean that comment as "do it or else..."; as I said, it was "minor nitpicking". Either way, thank you for making it consistent.

**Comment by est31** [comment]
See [discussion in #738](https://github.com/minetest/minetest_game/issues/738#issuecomment-172275279): I like the idea of the api and so on, but I don't like the new meshes, they have two door faces, the old only had one.

**Comment by sofar** [comment]
@est31 I just looked at ```backface_culling```, and it appears that ```mesh``` nodes don't support this, otherwise I would enable it.

**Comment by sofar** [comment]
@est31 I've updated this patch to enable backface culling as is added in this PR:

https://github.com/minetest/minetest/pull/3578

With that patch, the visual appearance is as you expect, and you do not see the double door.

**Comment by sofar** [comment]
![culling]
(http://i.imgur.com/qqMEVZa.png)

Notice the glass door is (intentionally) NOT culled, but the wooden door is.

**Comment by est31** [comment]
Really nice :)

**Comment by sofar** [comment]
Updated this branch with the necessary settings to enable backface_culling for the 3 doors that need it. This is dependent on minetest/minetest@3578 fix, which I think is now nailed down and ready for merging.

**Comment by sofar** [comment]
Rebased this as there was a small merge conflict due to 2 trapdoor changes being merged.

With the minetest backface_culling change for mesh and other nodes as of 2 days ago, this now can be cleanly applied and works as the screenshot shows.

**Comment by red-001** [comment]
Texture names need to be change to avoid unnecessary conflicts with other mods and texture packs.

**Comment by sofar** [comment]
@red-001 ack, will do.

**Comment by sofar** [comment]
Renamed all the textures to avoid texture pack conflict.

**Comment by paramat** [comment]
I need to test this but seems a good PR. One thing, can we have 'glass-edge textures' halfway down on both sides of the glass door like with the obsidian glass door? I'm thinking the bar is needed on both sides as the imagined handle would be located here. It also makes it clear the glass doors are made from 2 sections of glass, curently the glass door is in 2 parts on one side and 1 part on the other.

**Comment by paramat** [comment]
Testing, nice. The obsidian door has backface culling enabled, perhaps it should be the same as the glass door? Is it possible to make the 'dug' sounds of the glass doors the breaking glass sound?

**Comment by sofar** [comment]
@paramat the obsidian door texture could get some refining, and the idea of a double handle bar visual is probably a good idea. Maybe @red-001 wants to look at it?

**Comment by paramat** [comment]
I recommend adding the standard glass-edge texture halfway down the glass door where it's missing, then for both glass doors add a metallic (for a modern look) horizontal handle on one side, just as the wooden doors have. Maybe i'll do the texture myself ..

**Comment by paramat** [comment]
Here's an attempt, wide flat handles for a modern look:
doors_door_glass2
doors_door_obsidian_glass2

**Comment by sofar** [comment]
I like the ideas sofar, I'll take your textures and change the culling setting, and refresh this PR later today.

**Comment by sofar** [comment]
@paramat, taking your textures verbatim. Made obsidian door no longer backface_culled.

**Comment by red-001** [comment]
@paramat I did that but it look strange when you look at it from an angle but just be my option.

**Comment by red-001** [comment]
@sofar `door:hidden` should have an empty `on_blast` callback so TNT can't destroy it.

**Comment by sofar** [comment]
@red-001 good catch. I see that I had anticipated that, and written all the code to remove it in case the bottom door does get blasted and needs to be removed (a non-protected door), so this is a very simple change.

**Comment by sofar** [comment]
@red-001 what do you mean "inside". I'm not going to change the mesh model at this point. The whole idea is to keep it basic.

**Comment by paramat** [comment]
> I did that but it look strange when you look at it from an angle but just be my option.

Might be nice to remove the glass-edge horizontal from both sides so the doors are single sheets of glass, with the handle placed directly into the glass. Also thinking of adding hinges, will provide some alternative textures.

**Comment by paramat** [comment]
doors_door_glass3
doors_door_obsidian_glass3
These might look better, removed central divider, added hinges, moved handle closer to door edge because it is so wide.

**Comment by red-001** [comment]
@paramat yes that should look better.

**Comment by sofar** [comment]
Yeah, these are good. It's going to be fun watching people run into the obsidian doors :)

Let's not try to overdo this too much, though. We can always revisit these textures at a later point in time.

**Comment by sofar** [comment]
pushed the 2 new door textures from paramat.

**Comment by sofar** [comment]
FYI - here is the corresponding code changes needed for mesecons_doors:

https://github.com/Jeija/minetest-mod-mesecons/pull/249

@Jeija

**Comment by sofar** [comment]
I've pushed a minor update that updates the ```on_blast()``` methods to return an itemstack where this is needed. This will help a change in #806 to make TNT'd doors actually drop the doors as items. Otherwise doors are lost in the blast.

**Comment by paramat** [comment]
To avoid an ABM with interval5 for door conversion how about converting old doors to new on opening? if this is possible and fairly lightweight.

**Comment by sofar** [comment]
@paramat Not as simple as it seems, as I have to catch things like on_punch, screwdriver use, etc. It also likely breaks mesecons. 

The ABM is safer and more efficient: Once all the doors are converted it's never going to fire anymore.

**Comment by paramat** [comment]
Okay no problem.

**Comment by sofar** [comment]
@paramat with minetest/minetest@47464c9344a1a4817ad6e4c6ec44526df305b1f6 merged this can now be merged as well.

@ShadowNinja @sfan5, who else can +1 this so it can be merged?

**Comment by paramat** [comment]
Added line comments.

**Comment by Ekdohibs** [comment]
:+1:

File: mods/doors/init.lua [hunk 6]
```diff
@@ -1,435 +1,444 @@
...
+			end,
+			state = function(self)
+				local state = minetest.get_meta(self.pos):get_int("state")
+				if state % 2 == 1 then
+					-- open
+					return true
+				else
+					-- close
+					return false
+				end
```
Comment (kilbith) [inline mods/doors/init.lua]:
``` Lua
return state % 2 == 1
```

File: mods/doors/init.lua [hunk 7]
```diff
@@ -1,435 +1,444 @@
...
+			toggle = function(self, player)
+				return _doors.trapdoor_toggle(self.pos, player)
+			end,
+			state = function(self)
+				local name = minetest.get_node(pos).name
+				if string.sub(name, -5) == "_open" then
+					return true
+				else
+					return false
+				end
```
Comment (kilbith) [inline mods/doors/init.lua]:
``` Lua
return name:sub(-5) == "_open"
```

File: mods/doors/init.lua [hunk 8]
```diff
@@ -1,435 +1,444 @@
...
+	},
+}
+
+function _doors.door_toggle(pos, clicker)
+	local meta = minetest.get_meta(pos)
+	local state = meta:get_int("state")
+	local def = minetest.registered_nodes[minetest.get_node(pos).name]
+	local name = def.door.basename
+
+	if clicker ~= nil then
```
Comment (kilbith) [inline mods/doors/init.lua]:
``` Lua
if clicker then
```

File: mods/doors/init.lua [hunk 9]
```diff
@@ -1,435 +1,444 @@
...
-				local pn = placer:get_player_name()
-				local meta = minetest.get_meta(pt)
-				meta:set_string("doors_owner", pn)
-				meta:set_string("infotext", "Owned by "..pn)
-				meta = minetest.get_meta(pt2)
-				meta:set_string("doors_owner", pn)
-				meta:set_string("infotext", "Owned by "..pn)
+			if def.protected then
+				meta:set_string("doors_owner", placer:get_player_name())
+				meta:set_string("infotext", "Owned by " .. placer:get_player_name())
```
Comment (kilbith) [inline mods/doors/init.lua]: Please localize the placer name.

File: mods/doors/init.lua [hunk 10]
```diff
@@ -1,435 +1,444 @@
...
+
+function doors.register(name, def)
+	-- replace old doors of this type automatically
+	minetest.register_abm({
+		nodenames = {"doors:"..name.."_b_1", "doors:"..name.."_b_2"},
+		interval = 5.0,
+		chance = 1,
+		catch_up = true,
+		action = function(pos, node, active_object_count, active_object_count_wider)
+			local l = tonumber(string.sub(node.name, string.len(node.name)))
```
Comment (kilbith) [inline mods/doors/init.lua]:
``` Lua
local l = tonumber(node.name:match("(%d)$"))
```

Comment (sofar) [inline mods/doors/init.lua]: :match() is more expensive, it has to iterate over the string

File: mods/doors/init.lua (hunk 6 above)
Comment (sofar) [inline mods/doors/init.lua]: ack

File: mods/doors/init.lua (hunk 7 above)
Comment (sofar) [inline mods/doors/init.lua]: ack

File: mods/doors/init.lua (hunk 8 above)
Comment (sofar) [inline mods/doors/init.lua]: sure

File: mods/doors/init.lua (hunk 9 above)
Comment (sofar) [inline mods/doors/init.lua]: sure, yeah

**Comment by sofar** [comment]
I need to test the updates before I push, which will be later today.

File: mods/doors/init.lua (hunk 10 above)
Comment (kilbith) [inline mods/doors/init.lua]: Then use : 

``` Lua
local l = tonumber(node.name:sub(-1))
```

Comment (kaeza) [inline mods/doors/init.lua]: You can use -1 instead of `string.len`. Also, `node.name:sub(...)`, please. If it's really performance critical, localize.

Comment (sofar) [inline mods/doors/init.lua]: thanks

**Comment by sofar** [comment]
All tested well after the latest changes.

**Comment by paramat** [comment]
Seems well reviewed now so :+1: will merge soon.

**Comment by paramat** [comment]
f600a9f645af40d22c8eb7c17aff89507b71816e

**Comment by proller** [comment]
this pull brokes all mods using doors api.
protector (fixed)
landrush
and other

**Comment by sofar** [comment]
@proller I'll investigate, I had looked at homedecor and xdecor, but not those yet.

**Pull request closed without merging**